blog/part_3_option_parser.md

```diff
@@ -0,0 +1,201 @@
+
+  current_time =
+    current_date_time_in_time_zone
+    |> DateTime.truncate(:second)
+    |> DateTime.to_time
+    |> to_string()
+
+
+  assert time == current_time <> " Happy New Year Sydney!"
+end
+```
+
+Now we could think about testing this by travelling in time or creating a mock of the current time, or something else that sounds way to complex. But let's tackle this from a different perspective. First let's take a closer look at our implementation of the `run/1` function that takes in the timezone.
+
+```elixir
+def run([time_zone]) do
+  Mix.Task.run "app.start"
+
+  {:ok, current_date_time_in_time_zone} = DateTime.now(time_zone)
+
+  day_of_year =
+    current_date_time_in_time_zone
+    |> DateTime.to_date()
+    |> Date.day_of_year()
+
+  city =
+    String.split(time_zone, "/")
+    |> Enum.drop(1)
+    |> to_string()
+
+  time =
+    current_date_time_in_time_zone
+    |> DateTime.truncate(:second)
+    |> DateTime.to_time()
+    |> to_string()
+
+  output =
+    case day_of_year do
+      1 ->
+        "#{time} Happy New Year #{city}!"
+
+      _ ->
+        "time in #{city}: #{time}"
+    end
+
+  Mix.shell().info(output)
+end
+```
+
+Now our failing test fails because the output doesn't match the new year greeting we got on the 31st of december. We can make this very testable by exctracting just that part out in a different function and testing that in isolation.
+
+Let's create a new test for the new function we would like to see.
+
+```elixir
+test "the output contains a new years greeting on the first day of the year" do
+  time = ~T[02:00:00.005]
+  date = ~D[2019-01-01]
+  {:ok, date_time} = NaiveDateTime.new(date, time)
+
+  output = Mix.Tasks.Time.format_output(date_time, "Australia/Sydney")
+  truncated_time = time |> Time.truncate(:second) |> to_string()
+
+  assert output == truncated_time <> " Happy New Year Sydney!"
+end
+```
+
+I have hardcoded the timezone here for now, but that is good enough for now.
+We can make this test pass by creating the function extracting the output from the run function. We'll go ahead and extract the other conversions as private functions as well ending with the implementation below:
+
+```elixir
+def format_output(date_time, time_zone) do
+  time = time(date_time)
+  city = city(time_zone)
+
+  case day_of_the_year(date_time) do
+    1 ->
+      "#{time} Happy New Year #{city}!"
+
+    _ ->
+      "time in #{city}: #{time}"
+  end
+end
+
+defp city(time_zone) do
+  String.split(time_zone, "/")
+  |> Enum.drop(1)
+  |> to_string()
+end
+
+defp day_of_the_year(date_time) do
+  date_time
+  |> DateTime.to_date()
+  |> Date.day_of_year()
+end
+
+defp time(date_time) do
+  date_time
+  |> DateTime.to_naive
+  |> NaiveDateTime.truncate(:second)
+  |> NaiveDateTime.to_time()
+  |> to_string()
+end
+```
+
+Now that was a lot more refactoring than I initially thought I would do, but we pulled this out into separate functions very nicely!
```

Comment:
I agree this is a nice refactoring, but I'm not sure how it's relevant for the subject of this post?

Reply:
yes I agree. Maybe I can make the subject of this post New Years clean up? The OptionParser part is indeed the smallest part of the post.

What do you think?

Reply:
hmm or just name it refactoring, functional style and explain the pattern matching / arity matching of the functions?